### PR TITLE
 [Fire Mage] Initial spec setup + ABC

### PR DESCRIPTION
**Pull request by avilene**
Discussion can be found in #519 

- [x] Added basic spec
- [x] Added ABC support
- [x] Added T20 4set support
- [x] Add CastEfficiency

Next suggestions for modules:
* Add statistic for wasted casts when hot streak was up (cast anything but Pyro) and list which casts / how many
* Add statistic for overwritten Hot Streak procs (these two could be combined)
* Uptime on Pyretic Incantation
* Support for Meteor shoulders

### CODE REVIEW

File: src/Parser/FireMage/CHANGELOG.js
Comment (kfinch) [inline src/Parser/FireMage/CHANGELOG.js]: Changelog entries prefixed by class/spec, I think so they show right in overall changelog. Also attribution should be "(by Author)" Try  `10-14-2017 - Fire Mage: Added Spec. (by Fyruna)`

File: src/Parser/FireMage/CONFIG.js
Comment (kfinch) [inline src/Parser/FireMage/CONFIG.js]: Description not really intended to be about the dev, but about the analyzer. There's no one way to write them, but I think the RestoDruid config is a good example (of course I do, I wrote it 😄)

File: src/Parser/FireMage/Modules/Features/AlwaysBeCasting.js
Comment (kfinch) [inline src/Parser/FireMage/Modules/Features/AlwaysBeCasting.js]: Fix the string to be Fire Mage relevant. Probably all you need to do is swap in Scorch for Ice Lance.

File: src/Parser/FireMage/Modules/Items/ShardOfTheExodar.js
Comment (kfinch) [inline src/Parser/FireMage/Modules/Items/ShardOfTheExodar.js]: Shouldn't copypasta the entire Module. We need to come up for a plan for a shared location to put this so it's the same file for Frost and Fire (and eventually Arcane)

File: src/Parser/FireMage/Modules/Items/Tier20_4set.js
Comment (kfinch) [inline src/Parser/FireMage/Modules/Items/Tier20_4set.js]: Looks like this was left in by accident?

File: src/Parser/FireMage/Modules/Items/Tier20_4set.js
Comment (kfinch) [inline src/Parser/FireMage/Modules/Items/Tier20_4set.js]: Check events to make sure the damage event is consistently showing before the buff fades, sometimes WCL is weird about that.

File: src/Parser/FireMage/Modules/Items/Tier20_4set.js
Comment (kfinch) [inline src/Parser/FireMage/Modules/Items/Tier20_4set.js]: I recommend using `FIRE_MAGE_T20_4SET_BONUS_BUFF` for id / icon / title instead of Critical Massive, so it's more clear to the user what this is.

**Review by kfinch — CHANGES_REQUESTED** [review]
This is a good start, but it doesn't really include any suggestions yet. I believe Zerotorescue prefers that specs give several useful suggestions/statistics before we add them.

File: src/Parser/FireMage/Modules/Items/Tier20_4set.js
Comment (fyruna) [reply src/Parser/FireMage/Modules/Items/Tier20_4set.js]: How would I go about that? I'm not 100% sure on that.

File: src/Parser/FireMage/Modules/Items/ShardOfTheExodar.js
Comment (fyruna) [reply src/Parser/FireMage/Modules/Items/ShardOfTheExodar.js]: If we had the modules grouped by class and then spec it would have been easy enough to add a `Shared` folder for those modules. Shall I import the frost one for now?

File: src/Parser/FireMage/Modules/Items/Tier20_4set.js
Comment (kfinch) [reply src/Parser/FireMage/Modules/Items/Tier20_4set.js]: easiest would be to use the debug tools wowanalyzer has, like the event list. When running the analyzer locally using 'npm start', extra tabs will be available. Click the 'Events' tab, and use the filters to show what you're looking for.

Check that the damage event is happening before the buff fades.

File: src/Parser/FireMage/Modules/Items/ShardOfTheExodar.js
Comment (kfinch) [reply src/Parser/FireMage/Modules/Items/ShardOfTheExodar.js]: Leave it for now, we're discussing solutions in discord.

File: src/Parser/FireMage/Modules/Items/Tier20_4set.js
Comment (Sharrq) [reply src/Parser/FireMage/Modules/Items/Tier20_4set.js]: the event section towards the bottom would show it. Thats a good place for figuring out the order that stuff happens

File: src/Parser/FireMage/Modules/Features/AlwaysBeCasting.js
Comment (kfinch) [inline src/Parser/FireMage/Modules/Features/AlwaysBeCasting.js]: As Fire Blast isn't on the GCD, casting it while moving won't actually make any difference to your ABC stat.

**Comment by MartijnHols** [comment]
`This is a good start, but it doesn't really include any suggestions yet. I believe Zerotorescue prefers that specs give several useful suggestions/statistics before we add them.`

Actually I prefer a PR ASAP so that PRs stay small and reviewing is much easier. Especially with the completion ratings we have now adding just a skeleton spec should be fine.

**Comment by fyruna** [comment]
@MartijnHols then in that case this should be fine. I have an idea for a few suggestions/statistics to work on so when this is in I'll make that happen. (missed/unused procs and casts)

File: src/Parser/FireMage/CHANGELOG.js
Comment (sMteX) [reply src/Parser/FireMage/CHANGELOG.js]: Disagree - now that the changelogs are split by class there's no need to prefix it (it made sense when it was 1 Changelog for the whole analyzer.

**Comment by fyruna** [comment]
I think I'm missing CastEfficiency to be in a state where this is a barebones spec. Will add ASAP!

**Review by kfinch — DISMISSED** [review]

**Comment by kfinch** [comment]
Oh right this is missing CastEfficiency. Once that's in I'll merge.

**Comment by fyruna** [comment]
CastEfficiency works, I'll talk to the guild's fire mage to see what percentages would be ideal. (This is one of my own logs)
image

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: I think you also need to check for Soul of the Archmage too. (http://www.wowhead.com/item=151642/soul-of-the-archmage)

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: Should be `ROTATIONAL_AOE`

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: Should be `ROTATIONAL_AOE`

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: Should be `ROTATIONAL_AOE`

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: Should be `COOLDOWNS`

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: 15s CD

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: 15s CD

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: 24s CD

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: 5 minute CD

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (kfinch) [inline src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: Missing Scorch

**Review by kfinch — CHANGES_REQUESTED** [review]

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (fyruna) [reply src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: That grants you the actual talent so it's enough to check for whether you have it.

Comment (kfinch) [reply src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: To be sure, have you verified with your code against actual logs?

File: src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js
Comment (fyruna) [reply src/Parser/Mage/Fire/Modules/Features/CastEfficiency.js]: Added.

File: src/Parser/Mage/Fire/Modules/Features/CooldownTracker.js
Comment (sMteX) [inline src/Parser/Mage/Fire/Modules/Features/CooldownTracker.js]: I think I've discussed this with Sharrq. You basically have 2 options here. 

1) either leave Rune of Power in `cooldownSpells` in which case the `duration` is unnecessary. This will create a "cooldown window" every time you stand in Rune of Power. Therefore, if you leave and re-enter the Rune (removebuff and then applybuff again), it will create 2 "windows". 
2) if instead you put it in `castCooldowns`, you need to specify the `duration` and then it will create a fixed 10 second "window" on Rune of Power cast. The downside of this approach is that it tracks events that happen *outside* of the Rune of Power as well. 

So the choice is up to you what you want - `cooldownSpells` will be more accurate but potentially more messy, `castCooldowns` will be tidier but potentially less accurate.

**Review by sMteX — CHANGES_REQUESTED** [review]
I hope I didn't miss anything important - just that one thing you should decide, otherwise I think it's fine :)

File: src/Parser/Mage/Fire/Modules/Features/CooldownTracker.js
Comment (fyruna) [reply src/Parser/Mage/Fire/Modules/Features/CooldownTracker.js]: I'll be sticking with the cooldownSpells, it seems to be more accurate as if you are out of the rune we shouldn't be applying the buff to it.

**Review by sMteX — APPROVED** [review]

**Review by kfinch — APPROVED** [review]

**Pull request merged**
